src/main/python/deep_qa/layers/entailment_models/tuple_match.py

```diff
@@ -0,0 +1,180 @@
+        for layer in self.hidden_layers:
+            hidden_input = layer(hidden_input)
+        score_output = self.score_layer(hidden_input)
+        return score_output
+
+
+    def get_output_shape_for(self, input_shapes):
+        #pylint: disable=unused-argument
+        return (1,)
+
+
+    def build(self, input_shape):
+        # Build the dense layer(s)
+        self.hidden_layers = []
+        dense_input_shape = (None, self.num_entailment_features)
+        for i in range(self.num_hidden_layers):
+            curr_dense = Dense(output_dim=self.hidden_layer_width,
+                               activation=self.hidden_layer_activation,
+                               bias=self.use_biases,
+                               name='tuple_match_hidden_layer_%d' % i)
+            curr_dense.build(input_shape=dense_input_shape)
+            dense_input_shape = (None, self.hidden_layer_width)
+            self.hidden_layers.append(curr_dense)
+        score_layer_dense = Dense(output_dim=1,
+                                  activation='sigmoid',
+                                  name='tuple_match_score',
+                                  bias=self.use_biases)
+        score_layer_dense.build(input_shape=dense_input_shape)
+        self.score_layer = score_layer_dense
+
+        # Add the weights from these layers to TupleMatch's trainable weights
+        for hidden in self.hidden_layers + [self.score_layer]:
+            self.trainable_weights.extend(hidden.trainable_weights)
+
+        super(TupleMatch, self).build(input_shape)
+
+    @overrides
+    def compute_mask(self, inputs, input_mask=None):  # pylint: disable=unused-argument
+        # Here, input_mask is optionally a list of two masks (corresponding to the list of inputs) or None.
+        # Each of the masks in input_mask has shape: (batch size, num_slots)
+        # We want to see if either of the inputs is all padding (i.e. the mask would be all 0s), if so, then the
+        # whole tuple_match should be masked, so we would return a 0, otherwise we return a 1.  As such, the
+        # shape of the returned mask is (batch size, 1)
+        if any(input_mask):
+            mask1 = input_mask[0]
+            mask2 = input_mask[1]
+            mask1 = K.cast(K.any(mask1), 'int32')
+            mask2 = K.cast(K.any(mask2), 'int32')
+            return K.cast((mask1 + mask2) >= 2, 'int32')
+
+        return None
+
+    def call(self, inputs, mask=None):
+        tuple1_input = inputs[0]    # shape: (batch size, num_slots, num_slot_words_t1)
+        tuple2_input = inputs[1]    # shape: (batch size, num_slots, num_slot_words_t2)
+        # Check that the tuples have the same number of slots.
+        assert K.int_shape(tuple1_input)[1] == K.int_shape(tuple2_input)[1]
+
+        # Expand tuple1 to shape: (batch size, num_slots, num_slot_words_t1, num_slot_words_t2)
+        expanded_tuple1 = K.expand_dims(tuple1_input, 3)    # now (b, num_slots, num_slot_words_tuple1, __ )
```

Comment:
You can use `1` instead of `__` here, to match what you would see if you actually printed out `K.int_shape(tensor)`.